aip/client-libraries/4210.md

```diff
@@ -31,25 +35,95 @@ expectation.
+general GAPIC flow remains the same.
+
+```graphviz
+digraph {
+  node [ style="filled,solid" shape=box fontname="Roboto" ];
+  splines=ortho;
+  nodesep=0.3;
+  center=true;
+
+  proto [ label="API Protobuf\nDescriptors" shape=rectancle fillcolor=aliceblue ];
+
+  subgraph cluster_code_generators {
+    rank = same;
+    style = filled;
+    fillcolor = lightgrey;
+    node [ shape=oval ];
+
+    protobuf [ label="protobuf\ngenerator" fillcolor=deepskyblue3 ];
+    grpc [ label="gRPC\ngenerator" fillcolor=gold3 ];
+    gapic [ label="GAPIC\ngenerator" fillcolor=darkseagreen ];
+
+  }
+
+  proto -> protobuf;
+  proto -> grpc;
+  proto -> gapic;
+
+  subgraph cluster_generated_code {
+    rank = same;
+    style = filled;
+    fillcolor = lightgrey;
+    node [ shape=rectangle ];
+
+
+    protobuf_output [ label="Message & Enum\nCode" fillcolor=deepskyblue3 ];
+    grpc_output [ label="Server & Client\nStubs" fillcolor=gold3 ];
+    gapic_output [ label="Google API\nClient" fillcolor=darkseagreen ];
+
+  }
+  
+  protobuf -> protobuf_output;
+  grpc -> grpc_output;
+  gapic -> gapic_output;
+
+  assembly [ label="Package\nassembly" shape=oval fillcolor=aliceblue ];
+  
+  protobuf_output -> assembly
+  grpc_output -> assembly
+  gapic_output -> assembly
+
+  assembled_package [ label="Package of\ngenerated code" fillcolor=aliceblue ];
+
+  assembly -> assembled_package
+}
+```
+
+The following sections focus on the "GAPIC generator" in the above diagram.
+
 ### Protobuf plugins
 
-Code generators **must** be implemented as plugins to `protoc`, the protocol
-buffer compiler. The [plugin system][0] allows plugins to be written in any
-language, and plugins **should** ordinarily be written in the language being
-targeted, in order to take advantage of in-language tooling, and to ensure that
-experts in the target environment are able to meaningfully contribute.
+The protobuf compiler, `protoc`, supports a [plugin system][0] for code
+generation. The plugin system allows plugins to be written _in_ and _for_ any
+language.
+
+Code generators **must** be implemented as a `protoc` plugin. The following
+rules apply to the implementation of a client library generator as a `protoc`
+plugin:
 
+- The plugin  **should** be written in the language being targeted for
+  generation.
 - `protoc` expects plugins to be an executable in `$PATH`, and named
   `protoc-gen-{plugin_name}`, corresponding to the `--{plugin_name}_out` option
-  sent to the `protoc` executable.
-  - For a plugin creating client libraries for a specific language, the option
-    name **should** follow the convention `--{lang}_gapic_out` (meaning the
-    corresponding plugin executable is named `protoc-gen-{lang}_gapic`).
-- Plugins **must** accept a serialized `CodeGeneratorRequest` object (defined
+  sent to the `protoc` executable. As such:
+  - the plugin executable **should** be named `protoc-gen-{lang}_gapic`
```

Comment:
We definitely violate this for C#... I'd argue that it's actually better practice to explicitly specify the executable file when running a generator, so you know you're running the executable you expect rather than "some random binary which happens to be on the path".

Reply:
Yeah, I think PHP also violates it...I'm not sure I want to change a recommendation here regarding use of `--plugin` because that would impact our Bazel rule implementations (which invoke the generator binary) and would impact the plugin option names (thus all of the docs we have).

I think the main reasoning was that so all of them could be run at with a single `protoc` invocation unambiguously, but we all know that's not how our pipeline works today.

Can we put a pin this one? WDYT?

Reply:
Happy to punt on this for now, certainly.